truffle/docs/Safepoints.md

```diff
@@ -0,0 +1,178 @@
+This will allow, to collect guest language stack traces from other threads while they are blocked by uncooperative native code.
+Currently the safepoint will be invoked when the native code returns.
+
+## Tooling for Debugging
+
+There are several debug options available:
+
+### Excercise safepoints with SafepointALot 
+
+SafepointALot is a tool to exercise every safepoint of an application and collect statistics.
+
+If enabled with the `--engine.SafepointALot` option it prints the statistics on the cpu time interval between safepoints at the end of an execution.
+
+For example, running:
+
+```
+graalvm/bin/js --engine.SafepointALot js-benchmarks/harness.js -- octane-deltablue.js
+```
+
+Prints the following output to the log on context close:
+
+```
+DeltaBlue: 540
+[engine] Safepoint Statistics
+  --------------------------------------------------------------------------------------
+   Thread Name         Safepoints | Interval     Avg              Min              Max
+  --------------------------------------------------------------------------------------
+   main                  48384054 |            0.425 us           0.1 us       44281.1 us
+  -------------------------------------------------------------------------------------
+   All threads           48384054 |            0.425 us           0.1 us       42281.1 us
+```
+
+It is recommended for guest language implementations to try to stay below 1ms on average. 
+Note that precise timing can depend on CPU and interruptions by the GC.
+Since GC times are included in the safepoint interval times, it is expected that the maximum is close to the maximum GC interruption time.
+Future versions of this tool will be able to exclude GC interruption times from this statistic.
+
+### Trace thread local actions 
+
+The option `--engine.TraceThreadLocalActions` allows to trace all thread local actions of any origin.
+
+Example output:
+
+```
+[engine] [tl] submit                 0  thread[main]                     all-threads[alive=4]        side-effecting     asynchronous   action: com.oracle.truffle.api.test.TruffleSafepointTest$ActionCollector@6c779568
+[engine] [tl]   perform-start        0  thread[pool-1-thread-410]
+[engine] [tl]   perform-start        0  thread[pool-1-thread-413]
+[engine] [tl]   perform-start        0  thread[pool-1-thread-412]
+[engine] [tl]   perform-done         0  thread[pool-1-thread-413]
+[engine] [tl]   perform-done         0  thread[pool-1-thread-410]
+[engine] [tl]   perform-start        0  thread[pool-1-thread-411]
+[engine] [tl]   perform-done         0  thread[pool-1-thread-412]
+[engine] [tl]   perform-done         0  thread[pool-1-thread-411]
+[engine] [tl] done                   0  thread[pool-1-thread-411]
+```
+
+### Printing guest and host stack frames every time interval.
+
+The option `--engine.TraceStackTraceInterval=1000` allows to set the time in milliseconds to print the current stack trace on the next safepoint.
```

Comment:
As in it prints the stack on safepoints, but only once per `n` milliseconds?

Reply:
Clarified this a bit:

The option `--engine.TraceStackTraceInterval=1000` allows to set the time interval in milliseconds to repeatedly print the current stack trace.
Note that the stack trace is printed on the next safepoint poll and therefore might not be accurate.